pkg/provisioner/ironic/adopt_test.go

```diff
@@ -39,7 +39,7 @@ func TestAdopt(t *testing.T) {
 			ironic: testserver.NewIronic(t).Ready().WithNode(nodes.Node{
 				ProvisionState: string(nodes.Manageable),
 				UUID:           nodeUUID,
-			}).WithNodeStatesProvision(nodeUUID),
+			}).WithNodeStatesProvision(nodeUUID).WithNodeStatesProvisionUpdate(nodeUUID),
```

Comment:
Looking through this PR again, I think we want to be careful to not make the mock server always try to act exactly like Ironic. Most of the time we're going to want to keep the setup small and easy, since if we're only testing one part of the provisioner at a time.

So, I think we do want something like what you're adding here to support different methods for a node URL, but I don't think we want that to be the only way to set up the server and I don't think we want it to be the default. We haven't needed separate "update" calls until now. Let's see if we can extend the API in a way that doesn't require us to use them everywhere, or modify existing tests.

Maybe that means changing WithNode() and WithNodeStatesProvision() to register handlers for multiple methods? Tests that need separate methods could then call something other than WithNode() or WithNodeStatesProvision(), keeping the complication local to the test.

Reply:
On a second thought, I considered a mistake the fact that we didn't need the update calls until now: there was just a single handler registered for the same pattern that incidentally worked well for both the get/update cases.

From my point of view, the goal in a test case is to program the Ironic mock to return the desired values given some specific API calls, and performing clearly this declaration of intents in the the case could just make more readable the test itself.

I also agree on your point, and sometimes the underlying interaction operator <-> server can be pretty complex, and maybe in a test we're interested to "tweak" just few of the overall calls, relying on the default behavior for the others.
So, we could define a "default handler" (as we did initially for the catch all case), eventually configurable, to be used when the test author didn't set explicitly an handler (instead of failing as per my initial implementation).
At the risk of some magic, this anyhow will not force the author to define all the mock calls, but those ones just relevant for the given case, removing some of the complication.

Reply:
I like the idea of the default handler. It would help us balance the need to be explicit when configuring the mock server with the need to be concise in the tests to keep them readable.

Reply:
Ok, I've added the default handler and simplified the test code. I've added a new method `IronicMock.WithDefaultResponses()` so that the test author could enable the feature on a test-by-test case: it's not mandatory to describe every single call - but just those ones that are relevant for the test case.
The default handler accepts patterns with variables, and variables could be used in the payload to craft a little bit better the response (see MockServer.AddDefaultResponse()'.
Looks like that now with this approach we could be in the position to specify easily a default behavior, and override it on a case bases (and even for method) where needed.